### PR TITLE
Update build system for Tpetra support: CUDA and OpenMP

### PR DESCRIPTION
**Pull request by JulienLoiseau**
- Add Hypre GPU support using configure (instead of cmake)
- OpenMP support for Kokkos
- Update Hypre to 2.22.0

### CODE REVIEW

File: config/SuperBuild/include/Build_HYPRE.cmake [hunk 1]
```diff
@@ -22,57 +22,67 @@ amanzi_tpl_version_write(FILENAME ${TPL_VERSIONS_INCLUDE_FILE}
...
 endif()
 
-#find_package(LAPACK)
 if (LAPACK_FOUND)
-  set(hypre_lapack_opt "-DHYPRE_USING_HYPRE_LAPACK:BOOL=FALSE"
-                       "-DTPL_LAPACK_LIBRARIES:STRING=${LAPACK_LIBRARIES}")
+  if (CMAKE_HOST_SYSTEM_NAME STREQUAL "Darwin")
+    set(hypre_lapack_opt "--with-lapack-lib=-framework Accelerate")
+  else()
+    set(hypre_lapack_opt "--with-lapack-lib=${LAPACK_LIBRARIES}")
```
Comment (JulienLoiseau) [inline config/SuperBuild/include/Build_HYPRE.cmake:46]: This causes a problem with LAPACK_LIBRARIES=lib.so;-lm;-ldl

Comment (JulienLoiseau) [reply config/SuperBuild/include/Build_HYPRE.cmake:46]: @ethan I am not sure why this is adding the -lm and -ldl here and not in the flags.

Comment (JulienLoiseau) [reply config/SuperBuild/include/Build_HYPRE.cmake:46]: A solution can be to add: 
```
    string(REPLACE ";" "," L1 "${LAPACK_LIBRARIES}")
    set(hypre_lapack_opt "--with-lapack-lib=${L1}")
```

**Comment by lipnikov** [comment]
It would be useful for future support of this module is to leave a detailed comment why these two lines were removed.

File: config/SuperBuild/CMakeLists.txt [hunk 2]
```diff
@@ -222,6 +222,26 @@ if (ENABLE_Structured)
   endif()
 endif()
 
+message(STATUS ">>>>>> JDM: CMAKE_C_COMPILER_ID    = ${CMAKE_C_COMPILER_ID}")
```
Comment (lipnikov) [inline config/SuperBuild/CMakeLists.txt:225]: Why JDM?

File: config/SuperBuild/TPLVersions.cmake [hunk 3]
```diff
@@ -157,6 +157,7 @@
 #   0.97.12      - update Trilinos to afc4e525 (Nov 14, 2020)
 #   0.97.13      - update Alquimia to 1.0.8
 #   0.97.14      - update MSTK to 3.3.6
+#   0.97.15      - update Hypre to 2.22.0
```
Comment (lipnikov) [inline config/SuperBuild/TPLVersions.cmake:160]: To merge with master, this should be 0.98.1

File: config/SuperBuild/include/Build_HYPRE.cmake [hunk 4]
```diff
@@ -86,42 +96,50 @@ configure_file(${SuperBuild_TEMPLATE_FILES_DIR}/hypre-patch-step.cmake.in
...
+                         HYPRE_CUDA_SM=${Hypre_CUDA_SM}               
                     # -- Build
-                    BINARY_DIR       ${HYPRE_build_dir}        # Build directory 
-                    BUILD_COMMAND    ${MAKE} 
+                    BINARY_DIR       ${HYPRE_source_dir}/src        # Build directory 
+                    BUILD_COMMAND    $(MAKE)  
                     # -- Install
                     INSTALL_DIR      ${TPL_INSTALL_PREFIX}     # Install directory
-                INSTALL_COMMAND  $(MAKE) install
+      		          INSTALL_COMMAND  $(MAKE) install
```
Comment (lipnikov) [inline config/SuperBuild/include/Build_HYPRE.cmake:142]: Remove tabs.

File: src/solvers/test/solver_preconditioners.cc [hunk 5]
```diff
@@ -119,7 +119,7 @@ TEST(PRECONDITIONERS) {
   }
 };
 
-
+#if 0
```
Comment (lipnikov) [inline src/solvers/test/solver_preconditioners.cc:122]: Thy this test was removed?

File: tools/cmake/AmanziTPL.cmake [hunk 6]
```diff
@@ -284,7 +284,7 @@ if (Trilinos_FOUND)
       message(SEND_ERROR "Trilinos in ${Trilinos_DIR} does not have the Ifpack package")
     endif()
 
-    if (NOT Ifpack_ENABLE_HYPRE)
+    if (NOT Ifpack_ENABLE_HYPRE AND NOT ENABLE_KOKKOS)
```
Comment (lipnikov) [inline tools/cmake/AmanziTPL.cmake:287]: Does this remove Hypre if Kokkos is disabled?

**Review by lipnikov — APPROVED** [review]

**Comment by lipnikov** [comment]
This build re-compiles the code on the second loop and crashes in Alquimia. This did not happen before.

**Comment by jd-moulton** [comment]
Identified the problem with FindLAPACK.cmake adding -lm; -ldl to the list of libraries (see #578).  Fix will be put on master soon, and can be merged over here.

**Comment by jd-moulton** [comment]
RPATH for shared libraries is not set correctly in Silo, Hypre and Boost.  Ideally, we should fix these before merging.

**Pull request merged**